### ISSUE TITLE
Screenshots of Multiple Elements on Page Result in Some Blank Screenshots

### ISSUE THREAD

**Issue by juancarlosfarah** [puppeteer/puppeteer#2357]
### Steps to reproduce

**Tell us about your environment:**

* Puppeteer version: 1.2.0
* Platform / OS version: macOS 10.13.3
* URLs (if applicable): http://graasp.eu/ils/5a9d467e45aca55d33eafce6/?lang=en&printPreview
* Node.js version: 8.10.0

**What steps will reproduce the problem?**

I'm trying to take screenshots of a series of HTML elements. The first couple of elements in the series are captured without a problem. However,  the third through the tenth appear blank. The screenshots have the right dimensions, but they have no content. Interestingly, if I take a screenshot of the full page, the elements appear fine.

_Please include code that reproduces the issue._

```javascript
const puppeteer = require('puppeteer');

const captureScreenshotsOfElements = async (elements) => {
  let i = 0;
  for (const element of elements) {
    await element.screenshot({ path: `tmp/${i}.png` });
    i += 1;
  }
};

//
(async () => {
  const browser = await puppeteer.launch({headless: false});
  try {
    const page = await browser.newPage();
    await page.setViewport({width: 1200, height: 1200});
    await page.goto('http://graasp.eu/ils/5a9d467e45aca55d33eafce6/?lang=en&printPreview', { waitUntil: 'networkidle0' });

    // wait five more seconds just in case
    await page.waitFor(5000);

    // dismiss cookie banner
    const dismissCookiesMessageButton = 'a.cc-dismiss';
    await page.waitForSelector(dismissCookiesMessageButton);
    await page.click(dismissCookiesMessageButton);

    // print screenshot
    await page.screenshot({path: 'tmp/fullPage.png', fullPage: true});

    const gadgets = await page.$$('div.gadget-content');
    await captureScreenshotsOfElements(gadgets, page);

    await browser.close();
  } catch (err) {
    console.error(err);
    browser.close();
  }

})();
```

In short I do the following:

1. Load up the page.
2. Dismiss cookie banner.
3. Take screenshot of full page.
4. Select the elements and capture screenshots of each of them.

If have switched steps 3 and 4 around and get the same result.

**What is the expected result?**

All of the screenshots of the elements show the element content.

**What happens instead?**

The first three element screenshots show the content. The rest of them are empty.

I believe that there is a small moment after the viewport changes where the element is blank and that's when the screenshot is take. Is there anyway of waiting a bit before taking the screenshot of the element?

**Comment by joeldouglass** [comment]
I was able to work around this by setting the viewport height to a very large number (e.g. 1000000).

**Comment by juancarlosfarah** [comment]
@joeldouglass, I just tried doing that and it results in the browser crashing.

**Comment by michaelx** [comment]
**Env**
puppeteer-core 1.7.0
macOS 10.12.6
node.js 8.11.4

Having the same issue on a longer page. Increasing the viewport height was the only workaround that worked for me (waits didn’t change the behavior).

It does look like a common issue: https://github.com/GoogleChrome/puppeteer/issues/2423#issuecomment-383345790

@juancarlosfarah Have you tried setting the viewport height to the real page height _(you shouldn’t need more)_.

**Comment by eric-zeng** [comment]
I have also had issues with this, and none of the viewport workarounds that were posted on related issues worked out for me. It seems like the problem has something to do with how puppeteer/chrome resizes the viewport before clipping the element - I notice the page content going completely blank during the half second the screenshot is taken. 

However, I observed that when screenshotting the page, the page doesn't go blank. So I came up with workaround based on that: rather than using puppeteer to screenshot the element, I screenshot the whole page, and then outsource the cropping of the element to [sharp](https://github.com/lovell/sharp), an image processing library.

```javascript
const bb = await element.boundingBox();
if (!bb) {
    throw 'No bounding box';
}

// Sometimes the bounding box has non-integer values, in the sites I'm automating. Round the
// values just in case.
let boundingBox = {
    x: Math.floor(bb.x),
    y: Math.floor(bb.y),
    height: Math.ceil(bb.height),
    width: Math.ceil(bb.width)
}

// Take screenshot of page, then crop with sharp
const buf = await page.screenshot();
await sharp(buf).extract({
    left: boundingBox.x,
    top: boundingBox.y,
    width: boundingBox.width,
    height: boundingBox.height
}).toFile(pathname);

```
